private/bufpkg/bufprotoplugin/cmd/protoc-gen-go-apiclientconnect/main.go

```diff
@@ -159,14 +165,13 @@ func generatePackageFile(helper protogenutil.NamedHelper, plugin *protogen.Plugi
 		newClientGoIdent := connectGoImportPath.Ident("New" + interfaceName + "Client")
 		newClientGoIdentString := g.QualifiedGoIdent(newClientGoIdent)
 
+		g.P(`// New`, interfaceName, ` creates a new `, interfaceName)
 		g.P(`func (p *provider) New`, interfaceName, `(ctx `, contextGoIdentString, `, address string) (`, interfaceGoIdentString, `, error) {`)
-		g.P(`var contextModifier func(context.Context) context.Context`)
-		g.P(`var err error`)
-		g.P(`if p.contextModifierProvider != nil {`)
-		g.P(`contextModifier, err = p.contextModifierProvider(address)`)
-		g.P(`if err != nil {`)
-		g.P(`return  nil, err`)
-		g.P(`}`)
+
+		g.P(`interceptors := p.interceptors`)
+		g.P(`if p.authInterceptorProvider != nil {`)
+		g.P(`interceptor := p.authInterceptorProvider(address)`)
+		g.P(`interceptors = append(interceptors, interceptor)`)
```

Comment:
We're always attaching the auth interceptor to the very end - that means the request will propagate through the entire chain of interceptors before (potentially) failing at the end.

Today, the only other interceptor we configure is the CLI version header, so this is fine as-is. However, this could have unintended behavior depending on what other interceptors are configured before it. It should be fine to perform auth at the very end in general though.

If we're OK with this behavior, we just need to make it loud and clear that the auth interceptor will ALWAYS be applied last. That way if we ever care about the order, then we can make sure we change the order, or adapt as needed.

TL;DR Please leave a comment on the auth interceptor provider definition in the `bufcli` package (and potentially other places) that makes it clear where this interceptor takes effect in the chain.

Reply:
Updated.  Let me know if the comment makes it clear enough.

Reply:
After going through this ordeal, I'm beginning to realize that the provider pattern may not be the best approach.  It eliminates some boilerplate, but it also is a real headache in some corner cases.  The more of these corner cases I run into, the more I sour on the pattern.

What I may do in a future PR is just remove the ability to add `WithInterceptors` at all in the provider.  Right now, we only use one that we pass in, so it seems bug-prone for no good reason.